Let approx() work on more generic sequences

ndevenish opened the pull request:
### My Problem
approx() was updated in 9f3122fe to work better with numpy arrays, however at the same time the requirements were tightened from requiring an `Iterable` to requiring a `Sequence` - the former being
tested only on [interface](https://github.com/python/cpython/blob/f19447994983902aa88362d8fffe645f1ea2f2aa/Lib/_collections_abc.py#L252-L256), while the latter requires subclassing or [registration](https://github.com/python/cpython/blob/f19447994983902aa88362d8fffe645f1ea2f2aa/Lib/_collections_abc.py#L926-L929) with the abc.

This broke our code which uses many custom containers (though it wasn't apparent to us that things were broken until the Scalar testing was tightened in cd2085ee718fb297537638bb33a70fc781ea450f). For various (mainly political) reasons it's hard to change our code for pytest, and in any case I believe this benefits from a broader solution.

### This Patch
Since the ApproxSequence only used `__iter__` and `__len__` anyway, this patch reduces the requirement to only what's used, and allows unregistered Sequence-like containers to be used. Since numpy arrays qualify for the new criteria, I've reordered the checks so that generic sequences are checked for after numpy arrays (which would also work if numpy every made the arrays officially sequences)

### This Code

I know the new test doesn't look as neat as the old one, but thought it better to start out explicit rather than making e.g. a new custom `Sequencelike` abc which implements `__subclasshook__ `. `Sequence` can't be removed from `compat.py` because is also used for testing in https://github.com/pytest-dev/pytest/blob/176d27440ce04d26d9d6a2dd590a5f2cb39f4fa2/src/_pytest/assertion/util.py#L113-L114 and I haven't investigated whether it's also appropriate to loosen that. I'm happy to do that work if people would require.

I've categorised the `changelog` as `bugfix` for now. Also, have followed instructions but unsure if this is large enough to qualify for the step:
- Add yourself to `AUTHORS` in alphabetical order;

ndevenish commented:
Note: I realised I both left the AUTHORS change in, and got the order wrong, so fixed and re-pushed. Out of neatness habit I rewrote the main commit, so apologies for the extra work on the CI.

RonnyPfannschmidt commented on testing/python/approx.py:
```diff
@@ -496,3 +496,13 @@ def test_numpy_scalar_with_array(self):
+            def __len__(self):
+                return 4
+
+        assert [1, 2, 3, 4] == approx(newIterable())
```
i believe his breaks if an approx object is reused - just unpack explicitly

ndevenish replied:
Not entirely sure what you meant - unless you read the test class as an iterator rather than an iterable? In any case, I've split out the creations, was this what you meant?

https://github.com/pytest-dev/pytest/pull/4327/files#diff-a4d4d6cc8ecc32438c8a4da391852f22R508

ndevenish commented:
I've just been reminded that lack of `__iter__` doesn't mean that an object isn't iterable, as `iter` (and thus ranges, loops, comprehensions) will work just as well with `__getitem__` only.

So technically the test should be for` isinstance(expected, Iterable) or hasattr(expected, '__getitem__')`. Or, attempting `iter(expected)` and catching the exception?

Depends how general this should be?

nicoddemus commented:
Probably `isinstance(expected, Iterable)` is enough for most cases that we want to support. We may decide to be more general later if needed, but having to be more strict in the future is usually more problematic.

nicoddemus commented on testing/python/approx.py:
```diff
@@ -496,3 +496,13 @@ def test_numpy_scalar_with_array(self):
         assert approx(expected, rel=5e-8, abs=0) != actual
+
+    def test_generic_iterable_sized_object(self):
+        class newIterable(object):
```
I suggest using a `My` prefix (`MyIterable`) which is used in a few other places in our tests, and also follows the CamelCase naming convention for classes. 👍

nicoddemus commented on changelog/4327.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Loosen the definition of what ``approx`` considers a sequence
```
If we have decided this, please specify, I suggest:

```
``approx`` now works with more generic containers, more precisely instances of ``Iterable`` instead of more restrict ``Sequence``.
```

ndevenish replied:
I went for the slightly different

> ``approx`` again works with more generic containers, more precisely instances of ``Iterable`` and ``Sized`` instead of more restrictive ``Sequence``.

To emphasises the actual restriction and that it's restoring a previous compatibility

ndevenish commented:
Thanks for the comments. I've
- Changed the declaration of the expected iterable
- Changed the news fragment to be more precise
- Changed to be on top of the feature branch (though should I change the news classification to a feature also?

I think I wasn't supposed to force-push the branch to physically move it to the features branch, but now I know about that GitHub feature..

kalekundert commented:
It might be worth adding a test to make sure `__repr__()` works after making a comparison with a generator (i.e. because the generator would be consumed during the iteration, `approx` needs to make a copy in order to present a nice error message if the comparison fails).  This might just work—I don't remember how the code is written—but it seems worth testing.

nicoddemus commented:
> Changed to be on top of the feature branch (though should I change the news classification to a feature also?

Yes please. 👍 

> It might be worth adding a test to make sure __repr__() works after making a comparison with a generator

Oh good catch! It is definitely worth having a test for this. 👍

nicoddemus commented:
@RonnyPfannschmidt anything else here or can we merge this?
